From 91e0ab80a2e9689a5e7d541af0d586daab6ff110 Mon Sep 17 00:00:00 2001
From: Tianyi Hu <hutianyi@bytedance.com>
Date: Wed, 15 Sep 2021 21:43:18 +0800
Subject: [PATCH] DO NOT MERGE Hide overlay on KeyChainActivity

Hide non system overlay to improve security.

Test: N/A
Bug: 199754277
Merged-In: Ia0e97f40d79a7f89035572e0175990694870938f
Change-Id: Ia0e97f40d79a7f89035572e0175990694870938f
(cherry picked from commit 3de513868e45f022fce83d738032fc69b8c6b0f5)
---
 AndroidManifest.xml                            |  1 +
 src/com/android/keychain/KeyChainActivity.java | 10 ++++++++++
 2 files changed, 11 insertions(+)

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index 9d1601f..44f5463 100755
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -6,6 +6,7 @@
     <!-- Needed so KeyChainService on non-system user can write 
          security logging events -->
     <uses-permission android:name="android.permission.READ_LOGS"/>
+    <uses-permission android:name="android.permission.HIDE_NON_SYSTEM_OVERLAY_WINDOWS"/>
 
     <application android:label="@string/app_name"
             android:allowBackup="false"
diff --git a/src/com/android/keychain/KeyChainActivity.java b/src/com/android/keychain/KeyChainActivity.java
index 078e377..1279ab2 100644
--- a/src/com/android/keychain/KeyChainActivity.java
+++ b/src/com/android/keychain/KeyChainActivity.java
@@ -28,6 +28,7 @@ import android.content.pm.PackageManager;
 import android.content.res.Resources;
 import android.net.Uri;
 import android.os.AsyncTask;
+import android.os.Bundle;
 import android.os.RemoteException;
 import android.os.ServiceManager;
 import android.security.Credentials;
@@ -64,6 +65,8 @@ import java.util.stream.Collectors;
 
 import javax.security.auth.x500.X500Principal;
 
+import static android.view.WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS;
+
 public class KeyChainActivity extends Activity {
     private static final String TAG = "KeyChain";
 
@@ -81,6 +84,13 @@ public class KeyChainActivity extends Activity {
     // certificates.
     AlertDialog mLoadingDialog;
 
+
+    @Override
+    protected void onCreate(Bundle savedState) {
+        super.onCreate(savedState);
+        getWindow().addSystemFlags(SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS);
+    }
+
     @Override public void onResume() {
         super.onResume();
 
-- 
2.25.1

